src/overlays/actors/ovl_En_Holl/z_en_holl.c

```diff
@@ -24,37 +70,273 @@ const ActorInit En_Holl_InitVars = {
     (ActorFunc)EnHoll_Draw,
 };
 
-// static InitChainEntry sInitChain[] = {
-static InitChainEntry D_8089A5D0[] = {
+#include "overlays/ovl_En_Holl/ovl_En_Holl.c"
+
+static EnHoll* sInstancePlayingSound = NULL;
+
+static EnHollActionFunc sActionFuncs[] = {
+    EnHoll_VisibleIdle, EnHoll_VerticalIdle, EnHoll_TransparentIdle, EnHoll_VerticalBgCoverIdle, EnHoll_VisibleIdle,
+};
+
+static InitChainEntry sInitChain[] = {
     ICHAIN_F32(uncullZoneForward, 4000, ICHAIN_CONTINUE),
     ICHAIN_F32(uncullZoneScale, 400, ICHAIN_CONTINUE),
     ICHAIN_F32(uncullZoneDownward, 400, ICHAIN_STOP),
 };
 
-#endif
+//! Visible Halls: Being inside this plane enables the execution of more code within the actionFunc.
+static f32 sActivationPlaneDistance = 200.0f;
+
+//! Visible Halls: Crossing this plane will change rooms/scenes.
+static f32 sLoadingPlaneDistance = 150.0f;
+
+//! Visible Room Halls: Being inside this plane indicates the draw function will render using POLY_XLU.
+static f32 sTranslucencyPlaneDistance = 100.0f;
+
+//! Visible Room Halls: Being inside this plane indicates the draw function won't render.
+static f32 sTransparencyPlaneDistance = 50.0f;
+
+void EnHoll_SetupAction(EnHoll* this) {
+    this->type = EN_HOLL_GET_TYPE(this);
+    this->actionFunc = sActionFuncs[this->type];
+    if (EN_HOLL_IS_VISIBLE(this)) {
+        this->alpha = 255;
+    } else {
+        this->actor.draw = NULL;
+    }
+}
+
+void EnHoll_SetPlayerSide(GlobalContext* globalCtx, EnHoll* this, Vec3f* transformedPlayerPos) {
+    Player* player = PLAYER;
+
+    Actor_CalcOffsetOrientedToDrawRotation(&this->actor, transformedPlayerPos, &player->actor.world.pos);
+    this->playerSide = (transformedPlayerPos->z < 0.0f) ? EN_HOLL_BEHIND : EN_HOLL_BEFORE;
+}
+
+void EnHoll_Init(Actor* thisx, GlobalContext* globalCtx) {
+    EnHoll* this = THIS;
+    s32 pad;
+    Vec3f transformedPlayerPos;
+
+    Actor_ProcessInitChain(&this->actor, sInitChain);
+    EnHoll_SetupAction(this);
+    this->bgCoverAlphaActive = false;
+    this->alpha = 255;
+    EnHoll_SetPlayerSide(globalCtx, this, &transformedPlayerPos);
+}
+
+void EnHoll_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    EnHoll* this = THIS;
+
+    if (!EN_HOLL_IS_SCENE_CHANGER(this)) {
+        u32 enHollId = EN_HOLL_GET_ID_CAST(this);
+
+        globalCtx->doorCtx.transitionActorList[enHollId].id = -globalCtx->doorCtx.transitionActorList[enHollId].id;
+        if (this == sInstancePlayingSound) {
+            sInstancePlayingSound = NULL;
+        }
+    }
+}
+
+void EnHoll_ChangeRooms(GlobalContext* globalCtx) {
+    Room oldCurrRoom = globalCtx->roomCtx.currRoom;
+    globalCtx->roomCtx.currRoom = globalCtx->roomCtx.prevRoom;
+    globalCtx->roomCtx.prevRoom = oldCurrRoom;
+    globalCtx->roomCtx.activeMemPage ^= 1;
+}
```

Comment:
```suggestion
void EnHoll_ChangeRooms(GlobalContext* globalCtx) {
    Room prevRoom= globalCtx->roomCtx.currRoom;
    globalCtx->roomCtx.currRoom = globalCtx->roomCtx.prevRoom;
    globalCtx->roomCtx.prevRoom = oldCurrRoom;
    globalCtx->roomCtx.activeMemPage ^= 1;
}
```
isn't oldCurrRoom just prevRoom? Or would that be too confusing with `roomCtx.prevRoom`?

Reply:
Does the `SWAP` macro work here? i.e, `SWAP(Room, globalCtx->roomCtx.currRoom, globalCtx->roomCtx.prevRoom)`

Reply:
I was thinking that "prevRoom" would be confusing with roomCtx.prevRoom which is why I went with oldCurrRoom. Let me try the SWAP macro though, I wasn't aware of it.

Reply:
I'm getting two instructions reordered with the SWAP() macro and I'm not sure why. That's a bummer. I wonder if there's any way around that. I posted the diff in the server if anyone has any ideas.

Reply:
Apparently, it's because macros can't have line breaks, and line breaks affect codegen in IDO. RIP. It was a really good idea!

Reply:
Any preference on oldCurrRoom vs prevRoom though? I feel like oldCurrRoom illustrates that it is a temp better, and is also more descriptive of what it represents. Could also do tempRoom if folks don't like oldCurrRoom.